extensions/amp-apester-media/0.1/amp-apester-media.js

```diff
@@ -316,60 +329,67 @@ class AmpApesterMedia extends AMP.BaseElement {
-          .then(
-              response => {
-                const payload = response['payload'];
-                // If it's a playlist we choose a media randomly.
-                // The response will be an array.
-                const media = this.embedOptions_.playlist
-                  ? payload[Math.floor(Math.random() * payload.length)]
-                  : payload;
-                const src = this.constructUrlFromMedia_(media['interactionId']);
-                const iframe = this.constructIframe_(src);
-                const overflow = this.constructOverflow_();
-                const mutate = state => {
-                  state.element.classList.add('i-amphtml-apester-iframe-ready');
-                };
-                const state = {
-                  element: iframe,
-                  mutator: mutate,
-                };
-                this.mediaId_ = media['interactionId'];
-                this.iframe_ = iframe;
-                this.element.appendChild(overflow);
-                this.element.appendChild(iframe);
-                this.registerToApesterEvents_();
+    const vsync = Services.vsyncFor(this.win);
+    return this.queryMedia_().then(
+        response => {
+          const payload = response['payload'];
+          // If it's a playlist we choose a media randomly.
+          // The response will be an array.
+          const media = /** @type {JsonObject} */ (this.embedOptions_.playlist
+            ? payload[Math.floor(Math.random() * payload.length)]
+            : payload);
+          const src = this.constructUrlFromMedia_(media['interactionId']);
+          const iframe = this.constructIframe_(src);
+          this.intersectionObserverApi_ = new IntersectionObserverApi(
+              this,
+              iframe
+          );
 
-                return (this.iframePromise_ = this.loadPromise(iframe)
-                    .then(() => {
-                      Services.vsyncFor(this.win).runPromise({mutate}, state);
-                      return media;
-                    }));
-              },
-              error => {
-                dev().error(TAG, 'Display', error);
-                return undefined;
-              }
-          )
-    /** @param {!JsonObject} media */
-          .then(media => {
-            this.togglePlaceholder(false);
-            this.ready_ = true;
-            let height = 0;
-            if (media && media['data'] && media['data']['size']) {
-              height = media['data']['size']['height'];
-            }
-            if (height != this.height_) {
-              this.height_ = height;
-              if (this.random_) {
-                this./*OK*/ attemptChangeHeight(height);
-              } else {
-                this./*OK*/ changeHeight(height);
-              }
-            }
-          })
+          this.mediaId_ = media['interactionId'];
+          this.iframe_ = iframe;
+          this.registerToApesterEvents_();
+
+          return vsync.mutatePromise(() => {
```

Comment:
@OmriKeret per discussion below, this still needs a `then` that returns the whole chain, as it currently stands this mutatePromise resolves much sooner than loadPromise since the `return loadPromise` inside is ignored.

Reply:
Thanks @aghassemi, I didn't quite understand how the vsync.mutatePromise works before, I looked inside and saw that it doesn't resolve to any value, thus breaking the promise chain (which you have hinted already, but I didn't understand properly).
Anyway, I think that it's sorted out now. From my tests when the layout callback promise resolves than the unit is ready.